File: app/src/main/java/com/innov8/memeit/Fragments/ProfileFragment.java
package com.innov8.memeit.Fragments;


import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.tabs.TabLayout;
import com.innov8.memeit.Activities.TagsActivity;
import com.innov8.memeit.CustomClasses.CustomMethods;
import com.innov8.memeit.CustomClasses.ImageUtils;
import com.innov8.memeit.CustomClasses.UserListLoader;
import com.innov8.memeit.CustomViews.ProfileDraweeView;
import com.innov8.memeit.Fragments.ProfileFragments.UserListFragment;
import com.innov8.memeit.KUtilsKt;
import com.innov8.memeit.R;
import com.memeit.backend.MemeItUsers;
import com.memeit.backend.dataclasses.MyUser;
import com.memeit.backend.dataclasses.User;
import com.memeit.backend.utilis.OnCompleteListener;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;
import okhttp3.ResponseBody;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;


public class ProfileFragment extends Fragment implements Toolbar.OnMenuItemClickListener {
    private String userID;

    public static ProfileFragment newInstance(String uid) {
        ProfileFragment pf = new ProfileFragment();
        Bundle bundle = new Bundle();
        bundle.putString("uid", uid);
        pf.setArguments(bundle);
        return pf;
    }

    public static ProfileFragment newInstance(User user) {
        ProfileFragment pf = new ProfileFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable("user", user);
        bundle.putString("uid", user.getUserID());
        pf.setArguments(bundle);
        return pf;
    }

    public static ProfileFragment newInstance() {
        return new ProfileFragment();
    }

    public ProfileFragment() {

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            userID = getArguments().getString("uid", null);
            userData = getArguments().getParcelable("user");
        }
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.profile_page, container, false);
    }

    private TextView nameV;
    private TextView followerV;
    private TextView memeCountV;
    private ProfileDraweeView profileV;
    private TextView followBtnV;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        TabLayout tabLayout = view.findViewById(R.id.tabs_profile);
        ViewPager pager = view.findViewById(R.id.profile_viewpager);
        ViewPagerAdapter adapter = new ViewPagerAdapter(getChildFragmentManager());
        pager.setAdapter(adapter);
        tabLayout.setupWithViewPager(pager);
        nameV = view.findViewById(R.id.profile_name);
        Toolbar t = view.findViewById(R.id.toolbar);
        t.inflateMenu(R.menu.profile_page_menu);
        t.setOnMenuItemClickListener(this);
        followerV = view.findViewById(R.id.profile_followers_count);
        memeCountV = view.findViewById(R.id.profile_meme_count);
        profileV = view.findViewById(R.id.profile_image);

        followBtnV = view.findViewById(R.id.profile_follow_btn);
        followBtnV.setVisibility(isMe() ? GONE : VISIBLE);

        followBtnV.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (followBtnV.getText().toString().equalsIgnoreCase("Follow")) {
                    MemeItUsers.getInstance().followUser(userID, new OnCompleteListener<ResponseBody>() {
                        @Override
                        public void onSuccess(ResponseBody responseBody) {
                            Toast.makeText(getContext(), "Followed", Toast.LENGTH_SHORT).show();
                        }

                        @Override
                        public void onFailure(Error error) {
                            Toast.makeText(getContext(), "failed follow: " + error.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });
                } else if (followBtnV.getText().equals("Unfollow")) {
                    MemeItUsers.getInstance().unFollowUser(userID, new OnCompleteListener<ResponseBody>() {
                        @Override
                        public void onSuccess(ResponseBody responseBody) {
                            Toast.makeText(getContext(), "UnFollowed", Toast.LENGTH_SHORT).show();
                        }

                        @Override
                        public void onFailure(Error error) {
                            Toast.makeText(getContext(), "failed follow: " + error.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });

                }

            }
        });
        if (isMe()) loadFast();
        if (userData != null) updateView();
        loadData();
    }

    private void loadFast() {
        MyUser myUser = MemeItUsers.getInstance().getMyUser(getContext());
        nameV.setText(myUser.getName());
        profileV.setText(KUtilsKt.prefix(myUser.getName()));
        ImageUtils.loadImageFromCloudinaryTo(profileV, myUser.getImageUrl());

    }

    private User userData;

    private void loadData() {
        OnCompleteListener<User> onCompleteListener = new OnCompleteListener<User>() {
            @Override
            public void onSuccess(User user) {
                userData = user;
                updateView();
            }

            @Override
            public void onFailure(Error error) {
                Toast.makeText(getActivity(), "failer: " + error.getMessage(), Toast.LENGTH_SHORT).show();
            }
        };
        if (isMe())
            MemeItUsers.getInstance().getMyUserDetail(getContext(), onCompleteListener);
        else
            MemeItUsers.getInstance().getUserDetailFor(userID, onCompleteListener);

    }

    private void updateView() {
        nameV.setText(userData.getName());
        profileV.setText(KUtilsKt.prefix(userData.getName()));
        ImageUtils.loadImageFromCloudinaryTo(profileV, userData.getImageUrl());
        followerV.setText(CustomMethods.formatNumber(userData.getFollowerCount()));
        memeCountV.setText(CustomMethods.formatNumber(userData.getPostCount()));
        if (userData.isFollowedByMe()) {
            followBtnV.setText("Unfollow");
        } else {
            followBtnV.setText("Follow");
        }
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        switch (item.getItemId()){
            case R.id.menu_profile_tag:
                Intent intent=new Intent(getContext(), TagsActivity.class);
                if(!isMe()){
                    intent.putExtra("uid",userData.getUserID());
                }
                startActivity(intent);
                return true;

        }
        return false;
    }


    class ViewPagerAdapter extends FragmentPagerAdapter {
        String titles[] = {"Memes", "Following", "Followers"};
        Fragment[] fragments;

        ViewPagerAdapter(FragmentManager manager) {
            super(manager);
            fragments = new Fragment[]{MemeListFragment.newInstanceForUserPosts(userID),
                    UserListFragment.newInstance(UserListLoader.FOLLOWING_LOADER, userID),
                    UserListFragment.newInstance(UserListLoader.FOLLOWER_LOADER, userID)
            };
        }

        @Override
        public Fragment getItem(int position) {
            return fragments[position];
        }

        @Override
        public int getCount() {
            return fragments.length;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return titles[position];
        }
    }

    private boolean isMe() {
        String id = null;
        try {
            id = MemeItUsers.getInstance().getMyUser(getContext()).getUserID();
        } catch (NullPointerException ignored) {

        }
        return TextUtils.isEmpty(userID) || userID.equals(id);
    }
}

File: app/src/main/java/com/innov8/memeit/Fragments/ProfileFragment.kt
package com.innov8.memeit.Fragments


import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.ImageView
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import com.google.android.material.tabs.TabLayout
import com.innov8.memeit.Activities.ProfileSettingsActivity
import com.innov8.memeit.Activities.TagsActivity
import com.innov8.memeit.Activities.UserListActivity
import com.innov8.memeit.Activities.UserTagActivity
import com.innov8.memeit.Adapters.MemeAdapters.MemeAdapter
import com.innov8.memeit.CustomClasses.CustomMethods
import com.innov8.memeit.Loaders.FollowerLoader
import com.innov8.memeit.Loaders.FollowingLoader
import com.innov8.memeit.R
import com.innov8.memeit.Utils.loadImage
import com.innov8.memeit.Utils.prefix
import com.innov8.memeit.commons.toast
import com.memeit.backend.MemeItClient
import com.memeit.backend.MemeItUsers
import com.memeit.backend.call
import com.memeit.backend.models.User
import com.memeit.backend.models.UserReq
import kotlinx.android.synthetic.main.fragment_profile.*
import kotlinx.android.synthetic.main.loading_view_layout.*
import kotlinx.android.synthetic.main.profile_content.*


class ProfileFragment : Fragment(), Toolbar.OnMenuItemClickListener {
    private lateinit var userData: User

    private val isMe: Boolean
        get() = userData.uid == MemeItClient.myUser?.id

    private fun getMyUser() = MemeItClient.myUser!!
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        userData = arguments?.getParcelable("user")!!
        retainInstance = true
        initUser()
    }

    private fun initUser() {
        if (userData.uid == null) {
            if (userData.username != null) {
                if (userData.username == getMyUser().username) {
                    userData = getMyUser().toUser()
                    needToRefresh = true
                    loadedState = loaded
                    initViews()
                } else {
                    MemeItUsers.getUserByUserName(userData.username!!).call({
                        userData = it
                        needToRefresh = false
                        loadedState = loaded
                        initViews()
                    }) {
                        loadedState = failed
                        loader_view?.setError("Failed to load User Data")
                    }
                }
            }
        } else {
            if (userDataCompelete()) {
                needToRefresh = true
                loadedState = loaded
                initViews()
            } else {
                MemeItUsers.getUserById(userData.uid!!).call({
                    userData = it
                    needToRefresh = false
                    loadedState = loaded
                    initViews()
                }) {
                    loadedState = failed
                    loader_view?.setError("Failed to load User Data")
                }
            }
        }
    }

    private fun userDataCompelete() =
            userData.run {
                username != null && name != null
            }


    private fun refreshData() {
        val onLoaded = { user: User ->
            userData = user
            updateView()
        }
        val onError: (String) -> Unit = { context?.toast("Failed to refresh User Data") }

        if (isMe)
            MemeItUsers.getMyUser(onLoaded, onError)
        else
            MemeItUsers.getUserById(userData.uid!!).call(onLoaded, onError)


    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.loading_view_layout, container, false)
    }

    val loading = 0
    val loaded = 1
    val failed = 2

    var needToRefresh = true
    var loadedState = 0
        set(value) {
            field = value
            if (field == loaded && needToRefresh) {
                refreshData()
            }
        }

    fun initViews() {
        if (view == null) return
        if (loadedState != loaded) throw IllegalStateException("data should be loaded")
        loader_view.setLoaded()
        profile_viewpager.adapter = ViewPagerAdapter(childFragmentManager)
        tabs_profile.setupWithViewPager(profile_viewpager)
        listOf(
                R.drawable.ic_grid,
                R.drawable.ic_list,
                R.drawable.ic_badges
        ).forEachIndexed { index, id ->
            tabs_profile.getTabAt(index)?.apply {
                setCustomView(R.layout.tab_icon)
                (customView as ImageView).apply {
                    setImageResource(id)
                }
            }
        }

        initListeners()
        toolbar.inflateMenu(R.menu.profile_page_menu)
        if (!isMe) toolbar.menu.removeItem(R.id.menu_profile_id)
        profile_follow_btn.visibility = if (isMe) GONE else VISIBLE
        updateView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        loader_view.setContentView(R.layout.fragment_profile)
        loader_view.setErrorText("Failed to load user profile")
        loader_view.onRetry = { initUser() }
        if (loadedState == loaded)
            initViews()


    }

    private fun initListeners() {
        val onFollowerClicked = { _: View -> UserListActivity.startActivity(context!!, FollowerLoader(userData.uid)) }
        profile_followers.setOnClickListener(onFollowerClicked)
        profile_followers_count.setOnClickListener(onFollowerClicked)

        val onFollowingClicked = { _: View -> UserListActivity.startActivity(context!!, FollowingLoader(userData.uid)) }
        profile_followings.setOnClickListener(onFollowingClicked)
        profile_followings_count.setOnClickListener(onFollowingClicked)

        profile_follow_btn.setOnClickListener { _ ->
            if (profile_follow_btn.text == "Follow") {
                MemeItUsers.followUser(userData.uid!!).call({
                    profile_follow_btn.text = "Unfollow"
                    context?.toast("Followed")
                }, {
                    context?.toast("Failed to Follow:- $it")
                })
            } else if (profile_follow_btn.text == "Unfollow") {
                MemeItUsers.unfollowUser(userData.uid!!).call({
                    profile_follow_btn.text = "Follow"
                    context?.toast("Unfollowed")
                }, {
                    context?.toast("Failed to unfollow:- $it")
                })
            }
        }
        toolbar.setOnMenuItemClickListener(this)
        tabs_profile.addOnTabSelectedListener(object : TabLayout.BaseOnTabSelectedListener<TabLayout.Tab?> {
            override fun onTabReselected(p0: TabLayout.Tab?) {

            }

            override fun onTabUnselected(tab: TabLayout.Tab?) {
                (tab?.customView as ImageView).setColorFilter(Color.GRAY)

            }

            override fun onTabSelected(tab: TabLayout.Tab?) {
                (tab?.customView as ImageView).setColorFilter(Color.parseColor("#ff5e00"))
            }
        })
    }


    private fun updateView() {
        context?.run {
            profile_name?.text = userData.name
            profile_username?.text = userData.username
            profile_bio?.text = userData.bio
            profile_image?.setText(userData.name.prefix())
            profile_image?.loadImage(userData.imageUrl)
            profile_followers_count?.text = CustomMethods.formatNumber(userData.followerCount)
            profile_followings_count?.text = CustomMethods.formatNumber(userData.followingCount)
            profile_meme_count?.text = CustomMethods.formatNumber(userData.postCount)
            profile_follow_btn?.text = if (userData.isFollowedByMe) "Unfollow" else "Follow"
        }
    }


    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_profile_tag -> {
                if (isMe)
                    startActivity(Intent(context, TagsActivity::class.java))
                else {
                    val intent = Intent(context, UserTagActivity::class.java)
                    intent.putExtra("uid", userData.uid)
                    startActivity(intent)
                }
                return true
            }
            R.id.menu_profile_id -> {
                startActivity(Intent(context, ProfileSettingsActivity::class.java))
            }
        }
        return false
    }


    internal inner class ViewPagerAdapter(manager: FragmentManager) : FragmentPagerAdapter(manager) {
        //        var titles = arrayOf("Following", "Followers", "Memes", "Memes", "Badges")
        var fragments: Array<Fragment> = arrayOf(
                MemeListFragment.newInstanceForUserPosts(userData.uid!!,
                        if (isMe)
                            MemeAdapter.GRID_ADAPTER_MY_POSTS
                        else
                            MemeAdapter.GRID_ADAPTER_USER_POSTS),
                MemeListFragment.newInstanceForUserPosts(userData.uid!!,
                        if (isMe)
                            MemeAdapter.LIST_ADAPTER_MY_POSTS
                        else
                            MemeAdapter.LIST_ADAPTER_USER_POSTS),
                BadgeFragment.newInstance(if (isMe) null else userData.uid!!))

        override fun getItem(position: Int): Fragment = fragments[position]


        override fun getCount(): Int = fragments.size
        override fun getPageTitle(position: Int): CharSequence? = null


    }

    companion object {

        fun byID(uid: String): ProfileFragment {
            return ProfileFragment().apply {
                arguments = Bundle().apply {
                    putParcelable("user", User(uid = uid))
                }
            }
        }

        fun byUsername(username: String): ProfileFragment {
            return ProfileFragment().apply {
                arguments = Bundle().apply {
                    putParcelable("user", User(username = username))
                }
            }
        }

        fun byUser(user: User): ProfileFragment {
            return ProfileFragment().apply {
                arguments = Bundle().apply {
                    putParcelable("user", user)
                }
            }
        }

        fun newInstance(): ProfileFragment {
            return ProfileFragment().apply {
                arguments = Bundle().apply {
                    putParcelable("user", MemeItClient.myUser!!.toUser())
                }
            }
        }
    }
}
